src/Groups/group_constructors.jl

```diff
@@ -821,61 +821,107 @@ false
 Pc group of order 8
 
-julia> quaternion_group(PermGroup, 8)
+julia> dicyclic_group(PermGroup, 8)
 Permutation group of degree 8
 
-julia> g = quaternion_group(FPGroup, 8)
+julia> g = dicyclic_group(FPGroup, 8)
 Finitely presented group of order 8
 
 julia> relators(g)
 3-element Vector{FPGroupElem}:
  r^2*s^-2
  s^4
  r^-1*s*r*s
-
 ```
 """
-quaternion_group(n::IntegerUnion) = quaternion_group(PcGroup, n)
+dicyclic_group(n::IntegerUnion) = dicyclic_group(PcGroup, n)
 
-function quaternion_group(::Type{T}, n::IntegerUnion) where T <: GAPGroup
-  # FIXME: resolve naming: dicyclic vs (generalized) quaternion: only the
-  # former should be for any n divisible by 4; the latter only for powers of 2.
-  # see also debate on the GAP side (https://github.com/gap-system/gap/issues/2725)
+function dicyclic_group(::Type{T}, n::IntegerUnion) where {T<:GAPGroup}
   @assert iszero(mod(n, 4))
-  return T(GAP.Globals.QuaternionGroup(_gap_filter(T), n)::GapObj)
+  return T(GAP.Globals.DicyclicGroup(_gap_filter(T), n)::GapObj)
 end
 
 # Delegating to the GAP constructor via `_gap_filter` does not work here.
-function quaternion_group(::Type{T}, n::IntegerUnion) where T <: Union{PcGroup, SubPcGroup}
+function dicyclic_group(::Type{T}, n::IntegerUnion) where T <: Union{PcGroup, SubPcGroup}
   @assert iszero(mod(n, 4))
-  return T(GAP.Globals.QuaternionGroup(GAP.Globals.IsPcGroup, n)::GapObj)
+  return T(GAP.Globals.DicyclicGroup(GAP.Globals.IsPcGroup, n)::GapObj)
 end
 
 @doc raw"""
-    is_quaternion_group(G::GAPGroup)
+    is_dicyclic_group(G::GAPGroup)
 
-Return `true` if `G` is isomorphic to a (generalized) quaternion group
-of order $2^{k+1}, k \geq 2$, and `false` otherwise.
+Return `true` if `G` is isomorphic to a dicyclic group
+of order $4k, k > 1$, and `false` otherwise.
 
 # Examples
 ```jldoctest
-julia> is_quaternion_group(small_group(8, 3))
+julia> is_dicyclic_group(small_group(8, 3))
 false
 
-julia> is_quaternion_group(small_group(8, 4))
+julia> is_dicyclic_group(transitive_group(8, 5))
 true
 ```
 """
-@gapattribute is_quaternion_group(G::GAPGroup) = GAP.Globals.IsQuaternionGroup(GapObj(G))::Bool
+is_dicyclic_group(G::GAPGroup) = _is_dicyclic_group(G)
+
+@doc raw"""
+    is_quaternion_group(G::GAPGroup)
+
+Return `true` if `G` is isomorphic to a generalised quaternion group
+of order $2^{k+1}, k \geq 2$, and `false` otherwise.
+"""
+@gapattribute is_quaternion_group(G::GAPGroup) = GAP.Globals.IsGeneralisedQuaternionGroup(GapObj(G))
+
+# TODO: Remove once IsDicyclicGroup from GAP is available.
+function _is_dicyclic_group(G::GAPGroup)
+  N = order(G)
+  !iszero(mod(N, 4)) && return false
+  
+  n = N//2
+  a = n//2
+
+  G1, _ = derived_subgroup(G)
+  !(is_cyclic(G1) && order(G1) == a) && return false
+
+  local Zn
+  T = right_transversal(G, G1)
+  i = 1
+  while true
+    H = GAP.Globals.ClosureGroup(GapObj(G1), GapObj(T[i]))
+    Zn, _ = Oscar._as_subgroup(G1, H)
+    i = i + 1
+    if (i > 4) || (is_cyclic(Zn) && order(Zn) == n)
```

Comment:
I guess we could do away with `i` and just do this? But it would also make the code differ more from the GAP version, so perhaps not worth it anyway
```suggestion
  @assert length(T) == 4
  for x in T
    H = GAP.Globals.ClosureGroup(GapObj(G1), GapObj(x))
    Zn, _ = Oscar._as_subgroup(G1, H)
    if is_cyclic(Zn) && order(Zn) == n
```